### PR TITLE
Sticky bucketing (SDK)

### PR DESCRIPTION
**Pull request by bryce-fitzsimons**
### Features and Changes

- Closes **(add link to issue here)**

### CODE REVIEW

File: packages/sdk-js/src/GrowthBook.ts [hunk 1]
```diff
@@ -259,24 +265,56 @@ export class GrowthBook<
...
+
   public setAttributes(attributes: Attributes) {
     this._ctx.attributes = attributes;
-    if (this._ctx.remoteEval) {
-      this._refreshForRemoteEval();
-      return;
-    }
-    this._render();
-    this._updateAllAutoExperiments();
+    this.refreshStickyBuckets().then(() => {
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:298]: risky? setAttributes is no longer synchronous, even if sticky bucketing is disabled (side effects happen via `.then()` in the next loop cycle).

File: packages/sdk-js/src/GrowthBook.ts [hunk 2]
```diff
@@ -987,7 +1082,40 @@ export class GrowthBook<
...
+          )]: result.variationId,
+        }
+      );
+      if (changed) {
+        // update local docs
+        this._ctx.stickyBucketAssignmentDocs =
+          this._ctx.stickyBucketAssignmentDocs ?? {};
+        this._ctx.stickyBucketAssignmentDocs[attrKey] = doc;
+        // save doc
+        this._ctx.stickyBucketService?.saveAssignments(doc);
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:1116]: todo: determine if the `?.` operator causes bundle size inefficiencies during transpilation

Comment (jdorn) [reply packages/sdk-js/src/GrowthBook.ts:1116]: The `?.` operator technically checks if the value is either undefined or null, so the transpilation is pretty inefficient.  It will basically do:

```ts
(this._ctx.stickyBucketService === null || this._ctx.stickyBucketService === undefined) 
  ? undefined 
  : this._ctx.stickyBucketService.saveAssignments(doc)
```

Shorter to replace with:

```ts
this._ctx.stickyBucketService && this._ctx.stickyBucketService.saveAssignments(doc)
```

Not too bad when there's a single optional chain, but gets really verbose if you have multiple ones in a statement.

File: packages/sdk-js/src/GrowthBook.ts (hunk 1 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:298]: actually, any qualms with making this function async (since it's basically 90% async already without returning its promise)?

File: packages/sdk-js/src/GrowthBook.ts [hunk 3]
```diff
@@ -632,68 +674,95 @@ export class GrowthBook<
...
-        }
-        // If there are filters for who is included (e.g. namespaces)
-        if (rule.filters && this._isFilteredOut(rule.filters)) {
-          process.env.NODE_ENV !== "production" &&
-            this.log("Skip rule because of filters", {
-              id,
-              rule,
-            });
-          continue;
+        // Pre-check if we have an experiment-based sticky bucket, since that supersedes feature rules:
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:677]: @jdorn This is a carve out to make sure we are able to bypass attribute based targeting if a sticky bucket was found. A very similar lookup also happens in the experiment's _run() evaluation, but if we're ok with the approach, I can probably de-dupe the lookup. (Lookups are cached in the SDK anyhow so it's not super consequential)

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:694]: made this more abstract and changed how it affects specific rules. Still could use another set of eyes though

File: packages/sdk-js/src/GrowthBook.ts [hunk 4]
```diff
@@ -855,101 +954,133 @@ export class GrowthBook<
...
+      ) {
+        process.env.NODE_ENV !== "production" &&
+          this.log("Skip because of groups", {
+            id: key,
+          });
+        return this._getResult(experiment, -1, false, featureId);
+      }
+
+      // 8.2. Old style URL targeting
+      if (experiment.url && !this._urlIsValid(experiment.url as RegExp)) {
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:1013]: allow URL targeting for sticky buckets

File: packages/sdk-js/src/util.ts [hunk 5]
```diff
@@ -206,11 +207,22 @@ export function getBucketRanges(
     weights = equal;
   }
 
+  // If any variations are blocked, set their weight to 0 and rebalance
+  blockedVariations = blockedVariations || [];
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/util.ts:211]: for each bad range, apply the good ranges. don't alter the original weights

File: packages/sdk-js/src/types/growthbook.ts [hunk 6]
```diff
@@ -23,7 +23,12 @@ export type FeatureRule<T = any> = {
   weights?: number[];
   key?: string;
   hashAttribute?: string;
+  fallbackAttribute?: string;
   hashVersion?: number;
+  disableStickyBucketing?: boolean;
+  bucketVersion?: number;
+  minBucketVersion?: number;
+  excludeBlockedBucketUsers?: boolean;
```
Comment (jdorn) [inline packages/sdk-js/src/types/growthbook.ts:31]: What are `disableStickyBucketing` and `excludeBlockedBucketUsers` used for?

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/types/growthbook.ts:31]: - `disableStickyBucketing` is an experiment-level kill switch. We don't support it yet in the UI, but do in the SDK.
- `excludeBlockedBucketUsers` controls the behavior of sticky bucketed users (removes them instead of rebuckets) when we block a variation. Again there is no support in the UI for variation blocking yet.

Comment (jdorn) [reply packages/sdk-js/src/types/growthbook.ts:31]: I might get rid of the `excludeBlockedBucketUsers` part for now until we know for sure what data type will make sense when we do the UI side of things. For example, we might want to send down an array of blocked variations instead of just a  single boolean. Don't want to add things we aren't sure we'll actually need.

The kill switch, I'm fine leaving in if you want.

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/types/growthbook.ts]: sure, removed excludeBlockedBucketUsers. If the bucket variation key isn't found in the experiment, we now just re-bucket

File: docs/docs/lib/build-your-own.mdx [hunk 7]
```diff
@@ -1149,3 +1149,5 @@ Open a [GitHub issue](https://github.com/growthbook/growthbook/issues) with a li
   - Add 2 new test cases for matching on a `$groups` array attribute
 - **v0.5.2** 2023-10-30
   - Add 3 new test cases for comparison operators to handle more edge cases
+- **v0.5.3** 2023-10-30
```
[the remark this answers is not in the document]
Comment (bryce-fitzsimons) [reply docs/docs/lib/build-your-own.mdx]: - Addressed the `build-your-own` suggestions
- `cases.json` is tricky. I added a case for using `fallbackAttribute`, but testing the actual bucketing logic is tough since most of it requires consecutive E2E runs to ensure the bucketing persists. Open to suggestions here.

File: packages/sdk-js/src/GrowthBook.ts [hunk 8]
```diff
@@ -926,7 +1010,8 @@ export class GrowthBook<
       return this._getResult(experiment, -1, false, featureId);
     }
 
-    // 9. Get bucket ranges and choose variation
+    // 9. Get the variation from the sticky bucket or get bucket ranges and choose variation
+    let ranges: VariationRange[] = [];
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:1014]: Is this needed here?  Can we move this within the `if (!foundStickyBucket)` block below?

File: packages/sdk-js/src/GrowthBook.ts [hunk 9]
```diff
@@ -1149,4 +1297,151 @@ export class GrowthBook<
...
+      const attributes = this._getStickyBucketAttributes(data);
+      this.context.stickyBucketAssignmentDocs = await this.context.stickyBucketService.getAllAssignments(
+        attributes
+      );
+    }
+  }
+
+  private _getStickyBucketAssignments(): StickyAssignments {
+    const mergedAssignments: StickyAssignments = {};
+    Object.entries(this.context.stickyBucketAssignmentDocs ?? {}).forEach(
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:1342]: Can use `Object.values` instead if you don't need the key.

File: packages/sdk-js/src/GrowthBook.ts [hunk 10]
```diff
@@ -1149,4 +1297,151 @@ export class GrowthBook<
...
+      }
+    );
+    return mergedAssignments;
+  }
+
+  private _getStickyBucketVariation(
+    experimentKey: string,
+    experimentBucketVersion: number = 1,
+    minExperimentBucketVersion: number = 0,
+    meta: VariationMeta[] = []
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:1354]: Minor code golf: default attribute values tend to be very verbose when transpiled

File: packages/sdk-js/src/GrowthBook.ts [hunk 11]
```diff
@@ -1149,4 +1297,151 @@ export class GrowthBook<
...
+  ): {
+    variation: number;
+    versionIsBlocked: boolean;
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:1357]: Minor code golf: Make this field optional since `undefined` and `false` can generally be treated the same for comparisons

File: packages/sdk-js/src/GrowthBook.ts [hunk 12]
```diff
@@ -1149,4 +1297,151 @@ export class GrowthBook<
...
+  ): StickyExperimentKey {
+    return `${experimentKey}__${experimentBucketVersion}`;
+  }
+
+  private _getStickyBucketAttributes(
+    data?: FeatureApiResponse
+  ): Record<string, string> {
+    const attributes: Record<string, string> = {};
+    const shouldDeriveIdentifierAttributes =
+      data || this.context.stickyBucketIdentifierAttributes === undefined;
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:1407]: Does the `data ||` part need to be here?  That implies it will overwrite the explicitly set context options, but it feels like explicitly set options should always take precedence over derived attributes.

File: packages/sdk-js/src/GrowthBook.ts [hunk 13]
```diff
@@ -709,6 +755,12 @@ export class GrowthBook<
         if ("coverage" in rule) exp.coverage = rule.coverage;
         if (rule.weights) exp.weights = rule.weights;
         if (rule.hashAttribute) exp.hashAttribute = rule.hashAttribute;
+        if (rule.fallbackAttribute)
+          exp.fallbackAttribute = rule.fallbackAttribute;
+        if (rule.disableStickyBucketing)
+          exp.disableStickyBucketing = rule.disableStickyBucketing;
+        if (rule.bucketVersion) exp.bucketVersion = rule.bucketVersion;
```
[the remark this answers is not in the document]
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:762]: supposed to allow 0-indexing. I forgot to put a strict check here after changing from 1 to 0 in the UI.

File: packages/sdk-js/src/sticky-bucket-service.ts [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+      }
+    } catch (e) {
+      // Ignore localStorage errors
+    }
+    return doc;
+  }
+  async saveAssignments(doc: StickyAssignmentsDocument) {
+    const key = `${doc.attributeName}||${doc.attributeValue}`;
+    if (!this.localStorage) return;
+    await this.localStorage.setItem(this.prefix + key, JSON.stringify(doc));
```
Comment (jdorn) [inline packages/sdk-js/src/sticky-bucket-service.ts:120]: Should this be in a try/catch?

File: packages/sdk-js/src/sticky-bucket-service.ts [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+export class ExpressCookieStickyBucketService extends StickyBucketService {
+  /** intended to be used with cookieParser() middleware from npm: 'cookie-parser' **/
+  private prefix: string;
+  private req: RequestCompat | undefined;
+  private res: ResponseCompat | undefined;
+  private cookieAttributes: CookieAttributes;
+  constructor({
+    prefix = "gbStickyBuckets::",
```
Comment (jdorn) [inline packages/sdk-js/src/sticky-bucket-service.ts:131]: Are colons allowed in cookie names?  I see mixed results when I search online about browser support

File: packages/sdk-js/src/sticky-bucket-service.ts [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+    req,
+    res,
+    cookieAttributes = {},
+  }: {
+    prefix?: string;
+    req?: RequestCompat;
+    res?: ResponseCompat;
```
Comment (jdorn) [inline packages/sdk-js/src/sticky-bucket-service.ts:138]: For all of these, the arguments should be required so it can be caught easier at compile-time.  In this example, if `req` is not provided, it will throw an exception, but Typescript or eslint won't catch the bug.

File: packages/sdk-js/src/sticky-bucket-service.ts (hunk 15 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/sticky-bucket-service.ts]: changed to __ and also used encodeURIComponent (and decode...) for both cookie names & values when not already handled by whatever plugin each SB Service uses

**Review by jdorn — APPROVED** [review]
